File: android/app/src/main/java/com/cantstopthesignals/five/layout/TimeMap.java
package com.cantstopthesignals.five.layout;

import com.cantstopthesignals.five.Util;

import java.util.Calendar;
import java.util.List;

/**
 * Helper object to provide a map from times to pixels.
 */
public class TimeMap {
    private final List<Calendar> mTimeList;
    private final List<Integer> mYPosList;
    private final int mMillisPerDist;

    public TimeMap(List<Calendar> timeList, List<Integer> yPosList, int defaultDistancePerHour) {
        mTimeList = timeList;
        mYPosList = yPosList;
        mMillisPerDist = 60 * 60 * 1000 / defaultDistancePerHour;
        if (mTimeList.size() != mYPosList.size()) {
            throw new IllegalStateException("mTimeList and mYPosList should have same length");
        }
    }

    public int timeToYPos(Calendar time) {
        long timestamp = time.getTimeInMillis();
        int beforeIndex = -1;
        for (int i = 0, count = mTimeList.size(); i < count; i++) {
            if (time.compareTo(mTimeList.get(i)) <= 0) {
                break;
            }
            beforeIndex = i;
        }
        int afterIndex = beforeIndex + 1;
        if (beforeIndex < 0 || afterIndex >= mTimeList.size()) {
            if (beforeIndex != -1 && afterIndex != mTimeList.size()) {
                throw new IllegalStateException();
            }
            int anchorYPos;
            long anchorTimestamp;
            if (beforeIndex < 0) {
                // Timestamp is before first timestamp.
                anchorYPos = mYPosList.get(0);
                anchorTimestamp = mTimeList.get(0).getTimeInMillis();
            } else {
                // Timestamp is after first timestamp.
                anchorYPos = mYPosList.get(mYPosList.size() - 1);
                anchorTimestamp = mTimeList.get(mTimeList.size() - 1).getTimeInMillis();
            }
            return Util.round((timestamp - anchorTimestamp)
                    / mMillisPerDist) + anchorYPos;
        }
        int beforeYPos = mYPosList.get(beforeIndex);
        int afterYPos = mYPosList.get(afterIndex);
        long beforeTimestamp = mTimeList.get(beforeIndex).getTimeInMillis();
        long afterTimestamp = mTimeList.get(afterIndex).getTimeInMillis();
        return Util.round((timestamp - beforeTimestamp) *
                (afterYPos - beforeYPos) /
                (Math.max(1, afterTimestamp - beforeTimestamp))) + beforeYPos;
    }

    public Calendar yPosToTime(int yPos) {
        int beforeIndex = -1;
        for (int i = 0, count = mYPosList.size(); i < count; i++) {
            if (yPos <= mYPosList.get(i)) {
                break;
            }
            beforeIndex = i;
        }
        int afterIndex = beforeIndex + 1;
        if (beforeIndex < 0 || afterIndex >= mYPosList.size()) {
            int anchorYPos;
            long anchorTimestamp;
            if (beforeIndex != -1 && afterIndex != mYPosList.size()) {
                throw new IllegalStateException();
            }
            if (beforeIndex < 0) {
                // Position is before first position.
                anchorYPos = mYPosList.get(0);
                anchorTimestamp = mTimeList.get(0).getTimeInMillis();
            } else {
                // Position is after last position.
                anchorYPos = mYPosList.get(mYPosList.size() - 1);
                anchorTimestamp = mTimeList.get(mTimeList.size() - 1).getTimeInMillis();
            }
            long timestamp = Util.roundLong((yPos - anchorYPos) * mMillisPerDist) +
                    anchorTimestamp;
            Calendar time = Calendar.getInstance();
            time.setTimeInMillis(timestamp);
            return time;
        }
        int beforeYPos = mYPosList.get(beforeIndex);
        int afterYPos = mYPosList.get(afterIndex);
        long beforeTimestamp = mTimeList.get(beforeIndex).getTimeInMillis();
        long afterTimestamp = mTimeList.get(afterIndex).getTimeInMillis();
        long timestamp = Util.roundLong((yPos - beforeYPos) *
                (afterTimestamp - beforeTimestamp) /
                (Math.max(1, afterYPos - beforeYPos))) + beforeTimestamp;
        Calendar time = Calendar.getInstance();
        time.setTimeInMillis(timestamp);
        return time;
    }
}

File: web/src/main/javascript/layout/timemap.js
// Copyright cantstopthesignals@gmail.com

goog.provide('five.layout.TimeMap');

goog.require('five.util');
goog.require('goog.array');
goog.require('goog.asserts');
goog.require('goog.Disposable');
goog.require('goog.object');


/**
 * Helper object to provide a map from times to pixels.
 * @constructor
 * @extends {goog.Disposable}
 */
five.layout.TimeMap = function(timeList, yPosList,
    defaultDistancePerHour) {
  this.timeList_ = timeList;
  this.yPosList_ = yPosList;
  this.msPerDist_ = 60 * 60 * 1000 / defaultDistancePerHour;
  this.checkLists_();
};
goog.inherits(five.layout.TimeMap, goog.Disposable);

/** @param {goog.date.DateTime} time */
five.layout.TimeMap.prototype.timeToYPos = function(time) {
  var timestamp = time.getTime();
  var beforeIndex = -goog.array.binarySelect(this.timeList_,
      function(candidateTime) {
    return goog.date.Date.compare(goog.asserts.assertObject(time),
        candidateTime) || 1;
  }) - 2;
  var afterIndex = beforeIndex + 1;
  if (beforeIndex < 0 || afterIndex >= this.timeList_.length) {
    goog.asserts.assert(beforeIndex == -1 ||
        afterIndex == this.timeList_.length);
    if (beforeIndex < 0) {
      // Timestamp is before first timestamp.
      var anchorYPos = this.yPosList_[0];
      var anchorTimestamp = this.timeList_[0].getTime();
    } else {
      // Timestamp is after first timestamp.
      var anchorYPos = this.yPosList_[this.yPosList_.length - 1];
      var anchorTimestamp = this.timeList_[this.timeList_.length - 1].getTime();
    }
    var yPos = five.util.round((timestamp - anchorTimestamp)
        / this.msPerDist_) + anchorYPos;
    return yPos;
  }
  var beforeYPos = this.yPosList_[beforeIndex];
  var afterYPos = this.yPosList_[afterIndex];
  var beforeTimestamp = this.timeList_[beforeIndex].getTime();
  var afterTimestamp = this.timeList_[afterIndex].getTime();
  var yPos = five.util.round((timestamp - beforeTimestamp) *
      (afterYPos - beforeYPos) /
      (Math.max(1, afterTimestamp - beforeTimestamp))) + beforeYPos;
  return yPos;
};

/** @param {number} yPos */
five.layout.TimeMap.prototype.yPosToTime = function(yPos) {
  var beforeIndex = -goog.array.binarySelect(this.yPosList_,
      function(candidateYPos) {
    return (yPos - candidateYPos) || 1;
  }) - 2;
  var afterIndex = beforeIndex + 1;
  if (beforeIndex < 0 || afterIndex >= this.yPosList_.length) {
    goog.asserts.assert(beforeIndex == -1 ||
        afterIndex == this.yPosList_.length);
    if (beforeIndex < 0) {
      // Position is before first position.
      var anchorYPos = this.yPosList_[0];
      var anchorTimestamp = this.timeList_[0].getTime();
    } else {
      // Position is after last position.
      var anchorYPos = this.yPosList_[this.yPosList_.length - 1];
      var anchorTimestamp = this.timeList_[this.timeList_.length - 1].getTime();
    }
    var timestamp = five.util.round((yPos - anchorYPos) * this.msPerDist_) +
        anchorTimestamp;
    return new goog.date.DateTime(new Date(timestamp));
  }
  var beforeYPos = this.yPosList_[beforeIndex];
  var afterYPos = this.yPosList_[afterIndex];
  var beforeTimestamp = this.timeList_[beforeIndex].getTime();
  var afterTimestamp = this.timeList_[afterIndex].getTime();
  var timestamp = five.util.round((yPos - beforeYPos) *
      (afterTimestamp - beforeTimestamp) /
      (Math.max(1, afterYPos - beforeYPos))) + beforeTimestamp;
  return new goog.date.DateTime(new Date(timestamp));
};

five.layout.TimeMap.prototype.checkLists_ = function() {
  goog.asserts.assert(this.timeList_.length == this.yPosList_.length);
  for (var i = 0; i < this.timeList_.length; i++) {
    goog.asserts.assert(this.timeList_[i] instanceof goog.date.DateTime);
    goog.asserts.assertNumber(this.yPosList_[i]);
  }
};
